File: fsharp/space-age/SpaceAge.fs
module SpaceAge

open FSharp.Data.UnitSystems.SI.UnitNames

[<Measure>]
type year

type Planet =
    | Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

let orbitalPeriod =
    function
    | Mercury -> 0.2408467<year>
    | Venus -> 0.61519726<year>
    | Earth -> 1.0<year>
    | Mars -> 1.8808158<year>
    | Jupiter -> 11.862615<year>
    | Saturn -> 29.447498<year>
    | Uranus -> 84.016846<year>
    | Neptune -> 164.79132<year>

let private earthYearInSeconds = 31557600.0<second/year>

let private toYearOnEarth (seconds: int64): float<year> =
    float seconds * 1.0<second> / earthYearInSeconds

let private toYearOnPlanet (planet: Planet) (year: float<year>): float =
    year / orbitalPeriod planet

let age (planet: Planet) (seconds: int64): float =
    seconds
    |> toYearOnEarth
    |> toYearOnPlanet planet

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

yearInSeconds :: Float
yearInSeconds = 31557600

orbitalPeriod :: Planet -> Float
orbitalPeriod Mercury = 0.2408467
orbitalPeriod Venus = 0.61519726
orbitalPeriod Earth = 1.0
orbitalPeriod Mars = 1.8808158
orbitalPeriod Jupiter = 11.862615
orbitalPeriod Saturn = 29.447498
orbitalPeriod Uranus = 84.016846
orbitalPeriod Neptune = 164.79132

ageOn :: Planet -> Float -> Float
ageOn Mercury seconds = seconds / yearInSeconds / orbitalPeriod Mercury
ageOn Venus seconds = seconds / yearInSeconds / orbitalPeriod Venus
ageOn Earth seconds = seconds / yearInSeconds / orbitalPeriod Earth
ageOn Mars seconds = seconds / yearInSeconds / orbitalPeriod Mars
ageOn Jupiter seconds = seconds / yearInSeconds / orbitalPeriod Jupiter
ageOn Saturn seconds = seconds / yearInSeconds / orbitalPeriod Saturn
ageOn Uranus seconds = seconds / yearInSeconds / orbitalPeriod Uranus
ageOn Neptune seconds = seconds / yearInSeconds / orbitalPeriod Neptune

